### PR TITLE
Move tests to target .NET 8

### PR DESCRIPTION
**Pull request by Youssef1313**
Closes https://github.com/dotnet/roslyn-analyzers/issues/6490

### CODE REVIEW

**Comment by mavasani** [comment]
I think .NET 7 is preferable but tagging others to confirm their preference @buyaa-n @jmarolf @sharwell @stephentoub

**Comment by Youssef1313** [comment]
> I think .NET 7 is preferable but tagging others to confirm their preference @buyaa-n @jmarolf @sharwell @stephentoub

For reference, the .NET SDK got upgraded to .NET 8 by https://github.com/dotnet/roslyn-analyzers/pull/6409. The PR got created and merged automatically.

**Comment by Youssef1313** [comment]
Build is green. In case we want to keep on .NET 8, this should be ready to merge. Otherwise, I'll see if it's easy to get back to .NET 7 without build issues

**Comment by stephentoub** [comment]
> I think .NET 7 is preferable but tagging others to confirm their preference

Personally, I'd stick with .NET 7 for now until we're further along in .NET 8.

**Comment by Youssef1313** [comment]
This needs https://github.com/dotnet/roslyn-analyzers/pull/6441 to be merged first.

**Comment by Youssef1313** [comment]
@mavasani Can we merge this?

**Comment by mavasani** [comment]
@Youssef1313 checking out your branch locally and attempting to run unit tests in VS2022 gives the following error for every test:

```
Microsoft.CodeAnalysis.Analyzers.UnitTests.MetaAnalyzers.ReleaseTrackingAnalyzerTests.TestInvalidEntryDiagnosticAsync(entry: "Id1 | `<Undetected>` | Warning |", hasUndetectedField: True)
   Source: ReleaseTrackingAnalyzerTests.cs line 614
   Duration: 372 ms

  Message: 
NuGet.Resolver.NuGetResolverInputException : Unable to find package 'Microsoft.CodeAnalysis'. Existing packages must be restored before performing an install or update.

  Stack Trace: 
PackageResolver.Resolve(PackageResolverContext context, CancellationToken token)
ReferenceAssemblies.ResolveCoreAsync(String language, CancellationToken cancellationToken) line 272
ReferenceAssemblies.ResolveAsync(String language, CancellationToken cancellationToken) line 197
ReferenceAssemblies.ResolveAsync(String language, CancellationToken cancellationToken) line 174
AnalyzerTest`1.CreateSolutionAsync(ProjectId projectId, EvaluatedProjectState projectState, CancellationToken cancellationToken) line 1235
AnalyzerTest`1.CreateProjectImplAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, CancellationToken cancellationToken) line 1050
AnalyzerTest`1.CreateProjectAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, CancellationToken cancellationToken) line 1029
AnalyzerTest`1.GetSolutionAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, IVerifier verifier, CancellationToken cancellationToken) line 1007
AnalyzerTest`1.GetSortedDiagnosticsAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, ImmutableArray`1 analyzers, IVerifier verifier, CancellationToken cancellationToken) line 890
AnalyzerTest`1.VerifyDiagnosticsAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, DiagnosticResult[] expected, IVerifier verifier, CancellationToken cancellationToken) line 273
CodeFixTest`1.RunImplAsync(CancellationToken cancellationToken) line 304
AnalyzerTest`1.RunAsync(CancellationToken cancellationToken) line 181
ReleaseTrackingAnalyzerTests.VerifyCSharpAsync(String source, String shippedText, String unshippedText, DiagnosticResult[] expected) line 1021
ReleaseTrackingAnalyzerTests.TestInvalidEntryDiagnosticAsync(String entry, Boolean hasUndetectedField) line 638
--- End of stack trace from previous location ---
```

**Comment by Youssef1313** [comment]
@mavasani Does deleting `%TEMP%/test-packages` help?

**Comment by mavasani** [comment]
> @mavasani Does deleting `%TEMP%/test-packages` help?

Nope, same issue.

**Comment by Youssef1313** [comment]
@sharwell Any idea?

**Comment by mavasani** [comment]
@Youssef1313 I have identified the root cause for this issue. It's primarily due to the hardcoding of the Microsoft.CodeAnalysis package at below places: https://github.com/dotnet/roslyn-analyzers/blob/e51693353d557403d648d0d70226139f1b61dd60/src/Test.Utilities/AdditionalMetadataReferences.cs#L18

https://github.com/dotnet/roslyn-analyzers/blob/e51693353d557403d648d0d70226139f1b61dd60/src/Test.Utilities/AdditionalMetadataReferences.cs#L112

I am going to push the required fix to your branch. I'll also open a tracking issue for us to figure out some way to avoid this version hardcoding in source and instead derive this version from `MicrosoftCodeAnalysisVersion` property or at least have a comment in Versions.props to update these hardcoded values when `MicrosoftCodeAnalysisVersion` is bumped.

**Comment by Youssef1313** [comment]
Strange that this is only happening after moving to .NET 8.
I also don't understand why these versions have to match. What if I want to write a test for old version of MS.CA?

**Comment by mavasani** [comment]
> What if I want to write a test for old version of MS.CA?

Seems we actually do have tests which start failing when moved to newer version because they are explicitly relying on absence of new types added in later version of MS.CA. I am not sure how to go about fixing these - I am going to let these keep failing inside VS (as the tests still seem to pass from command line) and open a tracking issue to address this problem.

**Comment by mavasani** [comment]
@Youssef1313 So this causing almost half the tests to still fail inside VS. For example, all tests which use `ReferenceAssemblies.Net.Net70` are failing with:

```
System.Collections.Generic.KeyNotFoundException : The given key 'Microsoft.NETCore.App.Ref.7.0.0-preview.7.22375.6' was not present in the dictionary.

  Stack Trace: 
ImmutableDictionary`2.get_Item(TKey key)
ReferenceAssemblies.ResolveCoreAsync(String language, CancellationToken cancellationToken) line 341
ReferenceAssemblies.ResolveAsync(String language, CancellationToken cancellationToken) line 197
ReferenceAssemblies.ResolveAsync(String language, CancellationToken cancellationToken) line 174
AnalyzerTest`1.CreateSolutionAsync(ProjectId projectId, EvaluatedProjectState projectState, CancellationToken cancellationToken) line 1235
AnalyzerTest`1.CreateProjectImplAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, CancellationToken cancellationToken) line 1050
AnalyzerTest`1.CreateProjectAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, CancellationToken cancellationToken) line 1029
AnalyzerTest`1.GetSolutionAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, IVerifier verifier, CancellationToken cancellationToken) line 1007
AnalyzerTest`1.GetSortedDiagnosticsAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, ImmutableArray`1 analyzers, IVerifier verifier, CancellationToken cancellationToken) line 890
AnalyzerTest`1.VerifyDiagnosticsAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, DiagnosticResult[] expected, IVerifier verifier, CancellationToken cancellationToken) line 273
CodeFixTest`1.RunImplAsync(CancellationToken cancellationToken) line 304
AnalyzerTest`1.RunAsync(CancellationToken cancellationToken) line 181
ConstantExpectedTests.TestCSAsync(String source, DiagnosticResult[] diagnosticResults) line 1097
ConstantExpectedTests.TestArgumentConstantAsync(String type, String minValue, String maxValue, String value, String expression) line 586
--- End of stack trace from previous location ---
```

**Comment by mavasani** [comment]
@Youssef1313 Feel free to take a stab at this if you'd like. Seems like there many follow-up issues with the testing library to be resolved here.

Meanwhile, I would suggest that we remove the net472 target for test projects, as they fail inside VS and lead to confusion. Tests with netcoreapp31 target are passing fine inside VS as well as command line, so that should probably suffice.

**Comment by Youssef1313** [comment]
@mavasani Maybe this is related to sources we have in NuGet.config?

I think the test framework might be trying to restore from our root `NuGet.config` file. Feeds there might not be having packages with requested versions? Not sure if that is the case. Almost guessing.

I can take a closer look on the weekend.

**Comment by sharwell** [comment]
> We would need some way for test infrastructure to be able to restore the NuGet package with specified version.

It already does this (it's the only option it supports).

**Comment by Youssef1313** [comment]
@mavasani I rebased on latest `main` and all `Microsoft.CodeAnalysis.NetAnalyzers.UnitTests` tests pass in VS:

image

I haven't tried before the rebase though. Could you re-test?

**Comment by Youssef1313** [comment]
@mavasani @stephentoub Could we get this merged?

**Comment by Youssef1313** [comment]
Since the last commit is quite old, I'm closing and re-opening to start a new build, in case something went in between that fails the changes here.

File: eng/GenerateAnalyzerNuspec.targets [hunk 1]
```diff
@@ -119,7 +119,7 @@
   
   <PropertyGroup>
     <!-- Ideally, we would extract this from the MSBuild task, but we need this as the Target Output before that task is executed -->
-    <_GenerateAnalyzerNuspecPath>$(ArtifactsBinDir)GenerateAnalyzerNuspec\$(Configuration)\net7.0\GenerateAnalyzerNuspec.dll</_GenerateAnalyzerNuspecPath>
+    <_GenerateAnalyzerNuspecPath>$(ArtifactsBinDir)GenerateAnalyzerNuspec\$(Configuration)\$(NetCurrent)\GenerateAnalyzerNuspec.dll</_GenerateAnalyzerNuspecPath>
```
Comment (mavasani) [inline eng/GenerateAnalyzerNuspec.targets:122]: Where is this property $(NetCurrent) set?

Comment (Youssef1313) [reply eng/GenerateAnalyzerNuspec.targets:122]: @mavasani Introduced by Arcade in https://github.com/dotnet/arcade/pull/12161

It was advised that we use that https://github.com/dotnet/roslyn-analyzers/issues/6490

**Review by mavasani — APPROVED** [review]

**Pull request merged**